File: src/main/groovy/net/zomis/brainf/model/ast/tree/SyntaxTree.java
package net.zomis.brainf.model.ast.tree;

import net.zomis.brainf.model.BrainfuckRunner;

import java.util.ArrayList;
import java.util.List;
import java.util.ListIterator;

public class SyntaxTree extends Syntax implements Iterable<Syntax> {

    public List<Syntax> syntax = new ArrayList<>();

    @Override
    public void perform(BrainfuckRunner runner) {
        throw new IllegalStateException("Cannot perform a full syntax tree at once");
    }

    @Override
    public ListIterator<Syntax> iterator() {
        return syntax.listIterator();
    }

    @Override
    public String toString() {
        return "SyntaxTree{" +
                "syntax=" + syntax +
                '}';
    }

}

File: brainduck-kotlin/src/commonMain/kotlin/net/zomis/brainf/model/ast/tree/SyntaxTree.kt
package net.zomis.brainf.model.ast.tree;

import net.zomis.brainf.model.BrainfuckRunner

open class SyntaxTree : Syntax(), Iterable<Syntax> {

    val syntax = mutableListOf<Syntax>()

    override fun perform(runner: BrainfuckRunner) {
        throw IllegalStateException("Cannot perform a full syntax tree at once");
    }

    override fun iterator(): ListIterator<Syntax> {
        return syntax.listIterator()
    }

    fun iteratorAt(index: Int): ListIterator<Syntax> {
        return syntax.listIterator(index)
    }

    override fun toString(): String {
        return "SyntaxTree{" +
                "syntax=" + syntax +
                '}';
    }

}
